nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp

```diff
@@ -24,127 +24,137 @@
+
+    // Log a message every second
+    auto timer = node_->create_wall_timer(1s,
+        [&]() {RCLCPP_INFO(node_->get_logger(), "%s running...", primitive_name_.c_str());});
 
-    auto status = nav2_tasks::TaskStatus::FAILED;
+    loop(goal_handle);
 
+    stopRobot();
+  }
+
+  void initialCheck(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    if (onRun(goal_handle->get_goal()) != Status::SUCCEEDED) {
+      RCLCPP_INFO(node_->get_logger(), "Initial checks failed for ", primitive_name_.c_str());
+      goal_handle->abort(std::make_shared<typename ActionT::Result>());
+    }
+  }
+
+  void loop(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    auto result = std::make_shared<typename ActionT::Result>();
     rclcpp::Rate loop_rate(10);
+
     while (rclcpp::ok()) {
-      if (task_server_->cancelRequested()) {
-        RCLCPP_INFO(node_->get_logger(), "%s cancelled", taskName_.c_str());
-        task_server_->setCanceled();
-        status = nav2_tasks::TaskStatus::CANCELED;
-        break;
+      if (goal_handle->is_canceling()) {
+        RCLCPP_INFO(node_->get_logger(), "Canceling", primitive_name_.c_str());
+        goal_handle->canceled(result);
+        return;
       }
 
-      // Log a message every second
-      current_time = std::chrono::system_clock::now();
-      if (current_time - time_since_msg >= 1s) {
-        RCLCPP_INFO(node_->get_logger(), "%s running...", taskName_.c_str());
-        time_since_msg = std::chrono::system_clock::now();
-      }
+      // TODO(orduno) Handle or reject an attempted pre-emption
 
-      status = onCycleUpdate(result);
+      auto status = onCycleUpdate();
 
-      if (status == nav2_tasks::TaskStatus::SUCCEEDED) {
-        RCLCPP_INFO(node_->get_logger(), "%s completed successfully", taskName_.c_str());
-        break;
+      if (status == Status::SUCCEEDED) {
+        RCLCPP_INFO(node_->get_logger(), "%s completed successfully", primitive_name_.c_str());
+        // Primitives actions are empty msgs
+        goal_handle->succeed(result);
+        return;
       }
 
-      if (status == nav2_tasks::TaskStatus::FAILED) {
-        RCLCPP_WARN(node_->get_logger(), "%s was not completed", taskName_.c_str());
-        break;
+      if (status == Status::FAILED) {
+        RCLCPP_WARN(node_->get_logger(), "%s failed", primitive_name_.c_str());
+        goal_handle->abort(result);
+        return;
       }
 
-      if (status == nav2_tasks::TaskStatus::CANCELED) {
-        RCLCPP_WARN(node_->get_logger(), "%s onCycleUpdate() should not check for"
-          " task cancellation, it will be checked by the base class.", taskName_.c_str());
-        break;
-      }
       loop_rate.sleep();
     }
+  }
 
-    auto end_time = std::chrono::system_clock::now();
-    std::chrono::duration<double> elapsed_seconds = end_time - start_time;
-
-    RCLCPP_INFO(node_->get_logger(), "%s ran for %.2f seconds",
-      taskName_.c_str(), elapsed_seconds.count());
-
+  void stopRobot()
```

Comment:
With ROS2 Actions, I don't think we want to stop the robot this way. Instead, any currently executing action should be canceled. For example, using a failure in the BT.

Reply:
I dont think you should stop here because this is inside of `motion_primitives`. Its more than possible there's a primitive where that is nonsensical or potentially detrimental. I wouldn't assume how people will use this in their applications.

Reply:
@mjeronimo @SteveMacenski Before running a Motion Primitive we assume the robot has been stopped (a cancel for the `followPath` action has been issued and thus the dwb server stopped the robot). This method is used to stop after the primitive has completed (just in case). I'll assume it's up to the primitive to decide the last commanded velocity and remove this altogether to avoid confusion.

Reply:
Before running the *first* primitive, I agree with you, its been stopped, but then if you're running many of them in a row because you're trying to recover from some fault like delocalization, I can imagine a handful of situations where it might be adventitious for the robot to still be moving after the handoff to the next primitive to have a more dynamic recovery system of consecutive building blocks. 

I wouldn't add "stop" in the motion primitive base, but maybe at the beginning or end of the primitives you want to make sure you're stopped after.